README.md
# asciiCore
Simple particle simulation in ascii art.  
GitHub: https://github.com/YujiSODE/asciiCore  
Wiki: https://github.com/YujiSODE/asciiCore/wiki  
>Copyright (c) 2018 Yuji SODE \<yuji.sode@gmail.com\>  
>This software is released under the MIT License.  
>See LICENSE or http://opensource.org/licenses/mit-license.php
______

## 1. Description
It generates map list and displays this list.  
Command `::asciiCore::setObject` can define an object and returns object ID. Command `::asciiCore::remove` can remove one or more objects using object ID. Simulation settings can be reset by `::asciiCore::reset;`.  
Command `::asciiCore::run` runs simulation using defined objects. Map data can be cleared using command `::asciiCore::clear`. Simulated result can be output as a single JavaScript file using command `::asciiCore::output_JS`.

**`::asciiCore::clear;`**  
It clears map.

**`::asciiCore::setObject char rho ?xy0? ?v0? ?a0? ?cr?;`**  
It defines an object and returns its ID. 
- `char`: an object character
- `rho`: object density
- `xy0`: a list of initial object position with format of `{x y}`; default value is `{0.0 0.0}`.
- `v0`: a list of initial object velocities with format of `{vx vy}`; default value is `{0.0 0.0}`.
- `a0`: a list of initial object accelerations with format of `{ax ay}`; default value is `{0.0 0.0}`.
- `cr`: coefficient of restitution; default value is `1.0`.

**`::asciiCore::remove ?idList?;`**  
It removes objects with given ID list. All objects are removed if ID list is omitted. `idList` is a list of object IDs.

**`::asciiCore::reset;`**  
It resets size and scales for map, and removes all objects.

**`::asciiCore::run ?n? ?delay?;`**  
It runs simulation. `n` is model simulated for n-1 time steps, i.e., 0th step is the initial step. `delay`: delay in milliseconds with default value of `250`

**`::asciiCore::output_JS ?name?;`**  
It outputs log as JavaScript code. `name` is a name for output JavaScript function with default value of "asciiCore[numbers]"

For detailed description see [asciiCore_Documentation01](https://github.com/YujiSODE/asciiCore/wiki/asciiCore_Documentation01).

## 3. Script
#### Tcl
It requires Tcl 8.6+.  
- `asciiCore.tcl`: main script in Tcl
#### JavaScript
- `asciiCore_preview.js`: script to show simulated result in HTML

## 4. Library list
- Sode, Y. 2018. lSum.tcl: https://gist.github.com/YujiSODE/1f9a4e2729212691972b196a76ba9bd0
- Sode, Y. 2018. lPairwise.tcl: the MIT License; https://gist.github.com/YujiSODE/0d520f3e178894cd1f2fee407bbd3e16

## 5. Reference
- Tanaka, K. 2017. JavaScript game programming. Future Coders, Impress R&D. ISBN: 978-4-8443-9755-7


CHANGELOG.md
# Change Log

## [Unreleased]

## [v1.0 min] -2018-12-14
- a minimum version
## Removed
- removed `asciiCore_sample00.js`
- removed `asciiCore_sample01.js`
- removed `asciiCore_sample02.js`
- removed `asciiCore_webSample.html`
- removed `index.html`
- removed `test_asciiCore00.tcl`
- removed `test_asciiCore01.tcl`
- removed `test_asciiCore02.tcl`

## [1.0] -2018-12-14

## [0.1.1 beta] -2018-12-12
## Added
- `index.html`

## Released
## [0.1 beta] - 2018-12-11


test_asciiCore.tcl
source -encoding utf-8 ./asciiCore.tcl;
::asciiCore::resize 20 20;
#################
::asciiCore::setObject @ 1;
::asciiCore::setObject + 1 {1 0} {-1 0};
::asciiCore::setObject 1 1 {15 10} {-1 1};
::asciiCore::setObject 2 1 {10 15} {1 -1};
::asciiCore::setObject 3 1 {9 2} {-1 0};
#################
::asciiCore::show 3 1000;
after 250;
parray ::asciiCore::idMap;
::asciiCore::output_JS;


test_asciiCore00.tcl
#test_asciiCore00.tcl
source -encoding utf-8 ./asciiCore.tcl;
::asciiCore::resize 20 10;
::asciiCore::setEnvX wind -0.05;
::asciiCore::setEnvY gravity [expr {9.8*0.08}];
#################
::asciiCore::setFree [::asciiCore::setIsolated [::asciiCore::setObject @ 1 {3 3} {0 0}]];
::asciiCore::setObject A 1 {6 3} {-1 0};
::asciiCore::setObject B 1 {10 0} {0 1};
::asciiCore::setObject C 1 {10 4} {0 -0.5};
#################
::asciiCore::run 8 1000;
parray ::asciiCore::idMap;
::asciiCore::output_JS {asciiCore_sample00};


test_asciiCore01.tcl
#test_asciiCore01.tcl
source -encoding utf-8 ./asciiCore.tcl;
::asciiCore::resize 20 10;
::asciiCore::setEnvY gravity [expr {9.8*0.08}];
#################
::asciiCore::setObject A 1 {6 3};
::asciiCore::setObject B 1 {4 0};
#################
set i 0;
while {$i<10} {
	::asciiCore::setFree [::asciiCore::setIsolated [::asciiCore::setObject @ 10 [list $i 8]]];
	incr i 1;
};
#################
::asciiCore::run 10 1000;
parray ::asciiCore::idMap;
::asciiCore::output_JS {asciiCore_sample01};


asciiCore_preview.js
/*asciiCore
*asciiCore_preview.js
*===================================================================
*	Copyright (c) 2018 Yuji SODE <yuji.sode@gmail.com>
*
*	This software is released under the MIT License.
*	See LICENSE or http://opensource.org/licenses/mit-license.php
*===================================================================
* A simple tool for showing ascii art
*=== Synopsis ===
* `asciiCore_preview(asciiArt?,delay?);`
* this function shows an array of ascii art in <pre> HTML element
* that has id of "asciiCore", and it returns intervalID for `clearInterval(intervalID)`
*--- Parameters ---
* - asciiArt: an array of ascii art
* - delay: an optional delay in milliseconds with default value of 250
*/
//===================================================================
//this function shows an array of ascii art in <pre> HTML element that has id of "asciiCore", and it returns intervalID for `clearInterval(intervalID)`
function asciiCore_preview(asciiArt,delay){
	// - asciiArt: an array of ascii art
	// - delay: an optional delay in milliseconds with default value of 250
	var slf=window,n=asciiArt.length,pre=slf.document.getElementById('asciiCore'),i=1,T='';
	delay=!delay?250:delay;
	//initial value
	pre.textContent=asciiArt[0];
	T=slf.setInterval(()=>{pre.textContent=asciiArt[i%n],i+=1;},delay);
	return T;
}


asciiCore_sample00.js
var asciiCore_sample00=function(){return ["__________B_________\n____________________\n____________________\n___@__A_____________\n__________C_________\n____________________\n____________________\n____________________\n____________________\n____________________","____________________\n_________B__________\n____________________\n___@A_______________\n_________C__________\n____________________\n____________________\n____________________\n____________________\n____________________","____________________\n____________________\n____________________\n___@________________\n_________B__________\n___A_____C__________\n____________________\n____________________\n____________________\n____________________","____________________\n____________________\n____________________\n___@_____B__________\n____________________\n____________________\n____________________\n__A______C__________\n____________________\n____________________","____________________\n____________________\n____________________\n___@________________\n_________B__________\n____________________\n____________________\n____________________\n____________________\n____________________","____________________\n____________________\n____________________\n___@________________\n____________________\n_________B__________\n____________________\n____________________\n____________________\n____________________","____________________\n____________________\n____________________\n___@________________\n____________________\n____________________\n____________________\n________B___________\n____________________\n____________________","____________________\n____________________\n____________________\n___@________________\n____________________\n____________________\n____________________\n____________________\n____________________\n____________________"];};

asciiCore.tcl
#asciiCore
#asciiCore.tcl
##===================================================================
#	Copyright (c) 2018 Yuji SODE <yuji.sode@gmail.com>
#
#	This software is released under the MIT License.
##===================================================================
#Simple particle simulation in ascii art.
#=== Synopsis ===
# - `::asciiCore::clear;`
# - `::asciiCore::getMap;`
# - `::asciiCore::xCoord x;`
# - `::asciiCore::yCoord y;`
# - `::asciiCore::plot id;`
# - `::asciiCore::setEnvX name value;`
# - `::asciiCore::setEnvY name value;`
# - `::asciiCore::getEnvX;`
# - `::asciiCore::getEnvY;`
# - `::asciiCore::setObject char rho ?xy0? ?v0? ?a0? ?cr?;`
# - `::asciiCore::setIsolated id;`
# - `::asciiCore::unsetIsolated id;`
# - `::asciiCore::setFree id;`
# - `::asciiCore::unsetFree id;`
# - `::asciiCore::remove ?idList?;`
# - `::asciiCore::ifCollision id1 id2;`
# - `::asciiCore::getCollision id1 id2;`
# - `::asciiCore::step;`
# - `::asciiCore::resize w h;`
# - `::asciiCore::setScales dw dh;`
# - `::asciiCore::reset;`
# - `::asciiCore::objectInfo ?idList?;`
#------
# - `::asciiCore::run ?n? ?delay?;`
# - `::asciiCore::preview ?delay?;`
# - `::asciiCore::output_JS ?name?;`
#
#=== Description ===
#It generates map list and displays this list.
#
#=== [Reference] ===
# - Tanaka, K. 2017. JavaScript game programming. Future Coders, Impress R&D
#   ISBN: 978-4-8443-9755-7
##===================================================================
set auto_noexec 1;
package require Tcl 8.6;
#*** <namespace ::tcl::mathfunc> ***
#=== lSum.tcl (Yuji SODE, 2018): https://gist.github.com/YujiSODE/1f9a4e2729212691972b196a76ba9bd0 ===
#Additional mathematical functions for Tcl expressions that returns sum of given list
proc ::tcl::mathfunc::lSum {list} {namespace path {::tcl::mathop};set S 0.0;set R 0.0;set T 0.0;foreach e $list {set R [+ $R [expr double($e)]];set T $S;set S [+ $S $R];set T [+ $S [expr {-$T}]];set R [+ $R [expr {-$T}]];};return $S;};
#*** <namespace ::asciiCore> ***
namespace eval ::asciiCore {
	#### Variables ####
	#=== map list ===
	variable map {};
	#idMap is array of id lists in a particular coordinate cell
	variable idMap;array set idMap {};
	#e.g., map coordinates (n x m)
		#\0123...n
		#0____..._
		#1____..._
		#:____..._
		#m____..._
	#mW and mH are initial width and height of the map
	#default map size=10x10
	variable mW 10;
	variable mH 10;
	#initial unit scales for map
	variable dW 1.0;
	variable dH 1.0;
	#=== initial value for diameter ===
	variable D [expr {min($dW,$dH)}];
	#initial values for map
	set i 0;
	while {$i<$mH} {
		lappend map "_[string repeat "\x20_" [expr {$mW-1}]]";
		incr i 1;
	};
	#=== log ===
	variable LOG {};
	variable nStep 0;
	#=== epsilon ===
	variable epsilon 1.0;
	while {1.0+$epsilon!=1.0} {
		set epsilon [expr {$epsilon/2}];
	};
	#=== environmental accelerations ===
	variable EnvX;array set EnvX {};
	variable EnvY;array set EnvY {};
	#-------------------------------------------------------------------
	#=== variables used for collision ===
	variable nextVx1 0.0;
	variable nextVy1 0.0;
	variable nextVx2 0.0;
	variable nextVy2 0.0;
	#-------------------------------------------------------------------
	#=== array parameters for objects ===
	#-------------------------------------+
	#parameter format:`value(object_id)=v`|
	#-------------------------------------+
	#objects and their density and mass
	variable Char;array set Char {};
	variable Rho;array set Rho {};
	variable M;array set M {};
	#positions
	variable X;array set X {};
	variable Y;array set Y {};
	#velocities
	variable vX;array set vX {};
	variable vY;array set vY {};
	#accelerations
	variable aX;array set aX {};
	variable aY;array set aY {};
	#coefficient of restitution
	variable CR;array set CR {};
	#<<< additional objects attributes >>>
	#isolation from collision
	variable Iso;array set Iso {};
	#free from environmental accelerations
	variable Free;array set Free {};
	#### Procedures ####
	#=== lPairwise.tcl (Yuji SODE, 2018): the MIT License; https://gist.github.com/YujiSODE/0d520f3e178894cd1f2fee407bbd3e16 ===
	#It returns pairwise combination of given list
	proc lPairwise {list} {set n [llength $list];set i 1;set LIST {};while {$n>1} {set i 1;while {$i<$n} {lappend LIST [list [lindex $list 0] [lindex $list $i]];incr i 1;};set list [lrange $list 1 end];set n [llength $list];};return $LIST;};
	#-------------------------------------------------------------------
	#it clears map
	proc clear {} {
		variable map {};variable idMap;variable mW;variable mH;
		set i 0;
		foreach e [array names idMap] {
			array unset idMap $e;
		};
		#new values for $map
		while {$i<$mH} {
			lappend map "_[string repeat "\x20_" [expr {$mW-1}]]";
			incr i 1;
		};
	};
	#it returns the current map
	proc getMap {} {
		variable map;
		return [join [lmap e $map {list [join $e {}];}] \n];
	};
	#it returns horizontal coordinate
	proc xCoord {x} {
		variable dW;
		set x [expr {double($x)}];
		return [expr {wide(floor($x/$dW))}];
	};
	#it returns vertical coordinate
	proc yCoord {y} {
		variable dH;
		set y [expr {double($y)}];
		return [expr {wide(floor($y/$dH))}];
	};
	#it plots an object on map
	proc plot {id} {
		# - $id: object ID
		variable map;variable idMap;variable mW;variable mH;variable Char;variable X;variable Y;
		#map coordinates
		set x [::asciiCore::xCoord $X($id)];
		set y [::asciiCore::yCoord $Y($id)];
		expr {!($x<0)&&!($x>$mW-1)&&!($y<0)&&!($y>$mH-1)?[lappend idMap($x,$y) $id]:0};
		return [expr {!($x<0)&&!($x>$mW-1)&&!($y<0)&&!($y>$mH-1)?[lset map "$y $x" $Char($id)]:$map}];
	};
	#it set an environmental acceleration in the horizontal direction
	proc setEnvX {name value} {
		variable EnvX;
		set value [expr {double($value)}];
		array set EnvX "$name $value";
		return "$name $value";
	};
	#it set an environmental acceleration in the vertical direction
	proc setEnvY {name value} {
		variable EnvY;
		set value [expr {double($value)}];
		array set EnvY "$name $value";
		return "$name $value";
	};
	#it returns total of environmental acceleration in the horizontal direction
	proc getEnvX {} {
		variable EnvX;
		set E [array names EnvX];
		return [expr {[llength $E]>0?lSum([lmap e $E {list $EnvX($e);}]):0.0}];
	};
	#it returns total of environmental acceleration in the vertical direction
	proc getEnvY {} {
		variable EnvY;
		set E [array names EnvY];
		return [expr {[llength $E]>0?lSum([lmap e $E {list $EnvY($e);}]):0.0}];
	};
	#it defines an object and returns its ID
	proc setObject {char rho {xy0 {0.0 0.0}} {v0 {0.0 0.0}} {a0 {0.0 0.0}} {cr 1.0}} {
		# - $char: an object character
		# - $rho: object density
		# - $xy0: a list of initial object position with format of `{x y}`
		# - $v0: a list of initial object velocities with format of `{vx vy}`
		# - $a0: a list of initial object accelerations with format of `{ax ay}`
		# - $cr: coefficient of restitution
		variable Char;variable Rho;variable M;variable X;variable Y;variable vX;variable vY;variable aX;variable aY;variable CR;variable Iso;variable Free;variable dW;variable dH;
		#ID is an object id
		set ID [format %x [expr {round(rand()*10**10)}]];
		#V is approximated object volume
		set V [expr {double($dW*$dH)}];
		#=== array parameters for objects ===
		#------ objects and their density and mass ------
		array set Char "$ID $char";
		array set Rho "$ID [expr {double($rho)}]";
		array set M "$ID [expr {double($Rho($ID)*$V)}]";
		#------ position ------
		array set X "$ID [expr {double([lindex $xy0 0])}]";
		array set Y "$ID [expr {double([lindex $xy0 1])}]";
		#------ velocities and accelerations ------
		array set vX "$ID [expr {double([lindex $v0 0])}]";
		array set vY "$ID [expr {double([lindex $v0 1])}]";
		array set aX "$ID [expr {double([lindex $a0 0])}]";
		array set aY "$ID [expr {double([lindex $a0 1])}]";
		#------ coefficient of restitution ------
		array set CR "$ID [expr {double($cr)}]";
		#<<< additional objects attributes >>>: isolation from other objects
		array set Iso "$ID [expr {!1}]";
		#<<< additional objects attributes >>>: free from environmental accelerations
		array set Free "$ID [expr {!1}]";
		#=== initial plot of the object ===
		::asciiCore::plot $ID;
		return $ID;
	};
	#<<< additional objects attributes >>>
	#it makes given object isolated from collision
	proc setIsolated {id} {
		# - $id: object ID
		variable Iso;
		array set Iso "$id [expr {!0}]";
		return $id;
	};
	#<<< additional objects attributes >>>
	#it unsets isolation attribute of given object
	proc unsetIsolated {id} {
		# - $id: object ID
		variable Iso;
		array set Iso "$id [expr {!1}]";
		return $id;
	};
	#<<< additional objects attributes >>>
	#it makes given object free from environmental accelerations
	proc setFree {id} {
		# - $id: object ID
		variable Free;
		array set Free "$id [expr {!0}]";
		return $id;
	};
	#<<< additional objects attributes >>>
	#it unsets free attribute of given object
	proc unsetFree {id} {
		# - $id: object ID
		variable Free;
		array set Free "$id [expr {!1}]";
		return $id;
	};
	#it removes objects with given ID list
	#all objects are removed if ID list is omitted
	proc remove {{idList {}}} {
		# - $idList: a list of object ID
		variable Char;variable Rho;variable M;variable X;variable Y;variable vX;variable vY;variable aX;variable aY;variable CR;variable Iso;variable Free;
		set idList [expr {![llength $idList]?[array names Char]:$idList}];
		foreach e $idList {
			array unset Char $e;
			array unset M $e;
			array unset Rho $e;
			array unset X $e;
			array unset Y $e;
			array unset vX $e;
			array unset vY $e;
			array unset aX $e;
			array unset aY $e;
			array unset CR $e;
			array unset Iso $e;
			array unset Free $e;
		};
	};
	#it returns if there is collision
	#returned value is true when there is collision
	proc ifCollision {id1 id2} {
		# - $id1 and $id2: object IDs
		variable X;variable Y;variable D;
		set x1 $X($id1);set x2 $X($id2);
		set y1 $Y($id1);set y2 $Y($id2);
		set d [expr {lSum("$x1**2 $x2**2 $y1**2 $y2**2 -2*$x1*$x2 -2*$y1*$y2")}];
		return [expr {!(($d-$D**2)>0)}];
	};
	#it estimates velocity vectors after collision
	proc getCollision {id1 id2} {
		# - $id1 and $id2: object IDs
		variable M;variable X;variable Y;variable vX;variable vY;variable CR;variable Iso;variable D;variable epsilon;variable nextVx1;variable nextVy1;variable nextVx2;variable nextVy2;
		set x1 $X($id1);set x2 $X($id2);
		set y1 $Y($id1);set y2 $Y($id2);
		set vx1 $vX($id1);set vx2 $vX($id2);
		set vy1 $vY($id1);set vy2 $vY($id2);
		#-------------------------------------------------------------------
		#true when both velocities are 0: (vx,vy)=(0.0,0.0)
		set zeroVxy1 [expr {!double($vx1)&&!double($vy1)}];
		set zeroVxy2 [expr {!double($vx2)&&!double($vy2)}];
		set vx1 [expr {!!$vx1?$vx1:$epsilon}];
		set vy1 [expr {!!$vy1?$vy1:$epsilon}];
		set vx2 [expr {!!$vx2?$vx2:$epsilon}];
		set vy2 [expr {!!$vy2?$vy2:$epsilon}];
		#-------------------------------------------------------------------
		#mass
		set m1 $M($id1);set m2 $M($id2);
		#coefficient of restitution
		set cr [expr {lSum("$CR($id1) $CR($id2)")/2}];
		#distance between id1 and id2
		set d [expr {lSum("$x1**2 $x2**2 $y1**2 $y2**2 -2*$x1*$x2 -2*$y1*$y2")}];
		set d [expr {$d!=0.0?$d:$epsilon}];
		#-------------------------------------------------------------------
		#=== unit normal vectors ===
		set normalX1 [expr {lSum("$x2 -$x1")/$d}];
		set normalY1 [expr {lSum("$y2 -$y1")/$d}];
		set normalX2 [expr {lSum("$x1 -$x2")/$d}];
		set normalY2 [expr {lSum("$y1 -$y2")/$d}];
		#=== unit tangent vectors ===
		set tangentX1 $normalY1;
		set tangentY1 [expr {-$normalX1}];
		set tangentX2 $normalY2;
		set tangentY2 [expr {-$normalX2}];
		#=== inner products ===
		#--- unit normal vectors and velocities ---
		set dotNorm1 [expr {lSum("$normalX1*$vx1 $normalY1*$vy1")}];
		set dotNorm2 [expr {lSum("$normalX2*$vx2 $normalY2*$vy2")}];
		#--- unit tangent vectors and velocities ---
		set dotTan1 [expr {lSum("$tangentX1*$vx1 $tangentY1*$vy1")}];
		set dotTan2 [expr {lSum("$tangentX2*$vx2 $tangentY2*$vy2")}];
		#=== normal components ===
		set normX1 [expr {$normalX1*$dotNorm1}];
		set normY1 [expr {$normalY1*$dotNorm1}];
		set normX2 [expr {$normalX2*$dotNorm2}];
		set normY2 [expr {$normalY1*$dotNorm2}];
		#=== tangent components ===
		set tanX1 [expr {$tangentX1*$dotTan1}];
		set tanY1 [expr {$tangentY1*$dotTan1}];
		set tanX2 [expr {$tangentX2*$dotTan2}];
		set tanY2 [expr {$tangentY2*$dotTan2}];
		#=== next vectors ===
		set vx1_2 [expr {lSum("$normX2 $tanX1")}];
		set vy1_2 [expr {lSum("$normY2 $tanY1")}];
		set vx2_2 [expr {lSum("$normX1 $tanX2")}];
		set vy2_2 [expr {lSum("$normY1 $tanY2")}];
		#vector sizes
		set v1_2 [expr {sqrt(lSum("$vx1_2**2 $vy1_2**2"))}];
		set v2_2 [expr {sqrt(lSum("$vx2_2**2 $vy2_2**2"))}];
		#when vector size is 0
		set v1_2 [expr {!!$v1_2?$v1_2:$epsilon}];
		set v2_2 [expr {!!$v2_2?$v2_2:$epsilon}];
		#next unit vectors
		set unitVx1 [expr {$vx1_2/$v1_2}];
		set unitVy1 [expr {$vy1_2/$v1_2}];
		set unitVx2 [expr {$vx2_2/$v2_2}];
		set unitVy2 [expr {$vy2_2/$v2_2}];
		#-------------------------------------------------------------------
		#=== next velocities ===
		set v12X [expr {lSum("$vx1 -$vx2")}];
		set v12Y [expr {lSum("$vy1 -$vy2")}];
		set v21X [expr {lSum("$vx2 -$vx1")}];
		set v21Y [expr {lSum("$vy2 -$vy1")}];
		set m12 [expr {lSum("$m1 $m2")}];
		#when total mass is 0
		set m12 [expr {$m12!=0.0?$m12:$epsilon}];
		#--- next velocities: v1=>u1 and v2=>u2 ---
		#u1
		set uX1 [expr {lSum("$v21X*$m2*$cr $m1*$vx1 $m2*$vx2")/$m12}];
		set uY1 [expr {lSum("$v21Y*$m2*$cr $m1*$vy1 $m2*$vy2")/$m12}];
		#vector size
		set u1 [expr {lSum("$uX1**2 $uY1**2")}];
		#u2
		set uX2 [expr {lSum("$v12X*$m1*$cr $m1*$vx1 $m2*$vx2")/$m12}];
		set uY2 [expr {lSum("$v12Y*$m1*$cr $m1*$vy1 $m2*$vy2")/$m12}];
		#vector size
		set u2 [expr {lSum("$uX1**2 $uY1**2")}];
		#-------------------------------------------------------------------
		#=== next values ===
		#conditions are true when both velocities are 0: (vx,vy)=(0.0,0.0)
		set nextVx1 [expr {$zeroVxy1&&$zeroVxy2?0.0:$unitVx1*$u1}];
		set nextVy1 [expr {$zeroVxy1&&$zeroVxy2?0.0:$unitVy1*$u1}];
		set nextVx2 [expr {$zeroVxy1&&$zeroVxy2?0.0:$unitVx2*$u2}];
		set nextVy2 [expr {$zeroVxy1&&$zeroVxy2?0.0:$unitVy2*$u2}];
		#=== if objects have attribute for isolation ===
		set nextVx1 [expr {!!$Iso($id1)?$vX($id1):$nextVx1}];
		set nextVy1 [expr {!!$Iso($id1)?$vY($id1):$nextVy1}];
		set nextVx2 [expr {!!$Iso($id2)?$vX($id2):$nextVx2}];
		set nextVy2 [expr {!!$Iso($id2)?$vY($id2):$nextVy2}];
		#=== removing variables ===
		unset x1 x2 y1 y2 vx1 vx2 vy1 vy2;
		#-------------------------------------------------------------------
		unset m1 m2 cr d zeroVxy1 zeroVxy2;
		unset normalX1 normalY1 normalX2 normalY2;
		unset tangentX1 tangentY1 tangentX2 tangentY2;
		unset dotNorm1 dotNorm2 dotTan1 dotTan2;
		unset normX1 normY1 normX2 normY2;
		unset tanX1 tanY1 tanX2 tanY2;
		unset vx1_2 vy1_2 vx2_2 vy2_2 v1_2 v2_2 unitVx1 unitVy1 unitVx2 unitVy2;
		unset v12X v12Y v21X v21Y m12 uX1 uY1 u1 uX2 uY2 u2;
		#-------------------------------------------------------------------
		return "$nextVx1 $nextVy1 $nextVx2 $nextVy2";
	};
	#it calculates next step
	proc step {} {
		variable mW;variable mH;variable dW;variable dH;variable Char;variable M;variable X;variable Y;variable vX;variable vY;variable aX;variable aY;variable aX;variable aY;variable Iso;variable Free;variable D;variable nextVx1;variable nextVy1;variable nextVx2;variable nextVy2;variable EnvX;variable EnvY;
		#max width and height
		set maxW [expr {double($mW)*$dW}];
		set maxH [expr {double($mH)*$dH}];
		#list of all IDs
		set idList [array names Char];
		set IDs [::asciiCore::lPairwise $idList];
		set nId [llength $IDs];
		#environmental accelerations
		set xEnv [::asciiCore::getEnvX];
		set yEnv [::asciiCore::getEnvY];
		#--- conditions ---
		#$nId<1 => an element
		#1<$nId<3 => 2 elements
		#$nId>3 => 3 or more elements
		if {$nId>1&&$nId<3} {
			set ID1 [lindex $IDs 0];
			set ID2 [lindex $IDs 1];
			set col [::asciiCore::ifCollision $ID1 $ID2];
			expr {!!$col?[::asciiCore::getCollision $ID1 $ID2]:0};
			set vX($ID1) [expr {!!$col?$nextVx1:$vX($ID1)}];
			set vY($ID1) [expr {!!$col?$nextVy1:$vY($ID1)}];
			set vX($ID2) [expr {!!$col?$nextVx2:$vX($ID2)}];
			set vY($ID2) [expr {!!$col?$nextVy2:$vY($ID2)}];
		} elseif {$nId>3} {
			foreach e $IDs {
				set ID1 [lindex $e 0];
				set ID2 [lindex $e 1];
				set col [::asciiCore::ifCollision $ID1 $ID2];
				expr {!!$col?[::asciiCore::getCollision $ID1 $ID2]:0};
				set vX($ID1) [expr {!!$col?$nextVx1:$vX($ID1)}];
				set vY($ID1) [expr {!!$col?$nextVy1:$vY($ID1)}];
				set vX($ID2) [expr {!!$col?$nextVx2:$vX($ID2)}];
				set vY($ID2) [expr {!!$col?$nextVy2:$vY($ID2)}];
			};
		};
		#=== plotting and removing objects ===
		foreach e $idList {
			set vX($e) [expr {lSum("$vX($e) $aX($e)")}];
			set vY($e) [expr {lSum("$vY($e) $aY($e)")}];
			#--- environmental accelerations ---
			set vX($e) [expr {$Free($e)?$vX($e):lSum("$vX($e) $xEnv")}];
			set vY($e) [expr {$Free($e)?$vY($e):lSum("$vY($e) $yEnv")}];
			#--- coordinates ---
			set X($e) [expr {lSum("$X($e) $vX($e)")}];
			set Y($e) [expr {lSum("$Y($e) $vY($e)")}];
			expr {!($X($e)<0)&&!($X($e)>$mW-1)&&!($Y($e)<0)&&!($Y($e)>$mH-1)?[::asciiCore::plot $e]:[::asciiCore::remove $e]};
		};
		#=== removing variables ===
		unset maxW maxH idList IDs nId xEnv yEnv;
	};
	#it resizes map size
	proc resize {w h} {
		# - $w and $h are new width and height for the map
		variable map {};variable mW;variable mH;variable Char;variable X;variable Y;
		set re {^[1-9][0-9]*$};
		set i 0;
		set mW [expr {[regexp $re $w]?$w:$mW}];
		set mH [expr {[regexp $re $h]?$h:$mH}];
		#clearing map
		::asciiCore::clear;
		#replot objects on map
		foreach e [array names Char] {
			::asciiCore::plot $e;
		};
		return "$mW $mH";
	};
	#it sets new unit scales for map
	proc setScales {dw dh} {
		# - $dw and $dh: positive numbers for map scales
		variable dW;variable dH;variable Char;variable Rho;variable M;variable D;
		set dW [expr {double($dw>0?$dw:$dW)}];
		set dH [expr {double($dh>0?$dh:$dH)}];
		#V is approximated object volume
		set V [expr {double($dW*$dH)}];
		#new value for diameter
		set D [expr {min($dW,$dH)}];
		#clearing map
		::asciiCore::clear;
		#replot objects on map
		foreach e [array names Char] {
			#new value for mass
			set M($e) [expr {double($Rho($e)*$V)}];
			::asciiCore::plot $e;
		};
		return "$dW $dH";
	};
	#it resets size and scales for map, and removes all objects
	proc reset {} {
		variable LOG {};variable nStep 0;variable nextVx1 0.0;variable nextVy1 0.0;variable nextVx2 0.0;variable nextVy2 0.0;variable EnvX;variable EnvY;
		set nEnvX [array names EnvX];
		set nEnvY [array names EnvY];
		if {![llength $nEnvX]} {
			foreach e $nEnvX {array unset EnvX $e;};
		};
		if {![llength $nEnvY]} {
			foreach e $nEnvY {array unset EnvY $e;};
		};
		::asciiCore::remove;
		::asciiCore::resize 10 10;
		::asciiCore::setScales 1.0 1.0;
		::asciiCore::clear;
		unset nEnvX nEnvY;
	};
	#it returns object information
	#information of all objects are returned if ID list is omitted
	proc objectInfo {{idList {}}} {
		# - $idList: a list of object ID
		variable Char;variable Rho;variable M;variable X;variable Y;variable vX;variable vY;variable aX;variable aY;variable Iso;variable Free;
		set idList [expr {![llength $idList]?[array names Char]:$idList}];
		#info is information list
		set info [list [list ID Char Rho M X Y vX vY aX aY Isolation Free]];
		foreach e $idList {
			lappend info [list $e $Char($e) $Rho($e) $M($e) $X($e) $Y($e) $vX($e) $vY($e) $aX($e) $aY($e) $Iso($e) $Free($e)];
		};
		return $info;
	};
};
##===================================================================
#it runs simulation
proc ::asciiCore::run {{n 10} {delay 250}} {
	# - $n: model simulated for n-1 time steps, i.e., 0th step is the initial step
	# - $delay: delay in milliseconds with default value of 250
	variable LOG;variable nStep;variable dW;variable dH;variable epsilon;
	set i 0;
	set n [expr {$n>0?$n:10}];
	#Clearing display
	puts stdout "\u1b\[2J";
	while {$i<$n} {
		expr {$i!=0?[::asciiCore::clear]:0};
		expr {$i!=0?[::asciiCore::step]:0};
		lappend LOG [::asciiCore::getMap];
		puts -nonewline stderr "\u1b\[1;1H[lindex $LOG $nStep]";
		after $delay;
		incr i 1;
		incr nStep 1;
	};
	puts stdout "\n\#===================================================================";
	puts stdout "unit scales: ${dW} x ${dH}\nsteps: [expr {$nStep-1}]";
	puts stdout "epsilon: $epsilon";
};
#it shows simulation result
proc ::asciiCore::preview {{delay 250}} {
	# - $delay: delay in milliseconds with default value of 250
	variable LOG;variable epsilon;
	set n [llength $LOG];
	#Clearing display
	puts stdout "\u1b\[2J";
	foreach e $LOG {
		puts -nonewline stderr "\u1b\[1;1H$e";
		after $delay;
	};
	puts stdout "\n\#===================================================================";
	puts stdout "steps: [expr {$n-1}]";
	puts stdout "epsilon: $epsilon";
};
#it outputs log as JavaScript code
proc ::asciiCore::output_JS {{name {}}} {
	# - $name: a name for output JavaScript function with default value of "asciiCore[numbers]"
	variable LOG;
	set logJS {};
	set js {};
	#--- name ---
	set name [expr {[llength $name]>0?$name:"asciiCore[clock seconds]"}];
	set name [string map {\x20 _ \t _ . _ , _ \: _ \; _} $name];
	#--- JavaScript function ---
	set js "var\x20$name=function()";
	set logJS [join [lmap e $LOG {string map {\n \\n} "\"$e\"";}] ,];
	append js "\{return\x20\[$logJS\]\;\}\;";
	#--- output ---
	set C [open "$name.js" w];
	fconfigure $C -encoding utf-8;
	puts -nonewline $C $js;
	close $C;unset C;
	#=== removing variables ===
	unset logJS js;
	return "$name.js";
};
#*** License ***
#MIT License
#
#Copyright (c) 2018 Yuji Sode
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
